### ISSUE TITLE
Failed to connect ESP tunnel; using HTTPS instead.

### ISSUE THREAD

**Issue by surjikal** [dlenski/openconnect#75]
# Problem description

My VPN connection is slow and I'd like to see if an ESP tunnel will fix the speed issues. Unfortunately I can't get it to work, and so I need your help! SSL / HTTPS tunnels works. Great work btw!

Here's how I'm running openconnect:

```
echo $PASSWORD | openconnect --dump -v --protocol=gp --disable-ipv6 --csd-wrapper="hip.sh"  --non-inter --compression=all --user=surjikal --passwd-on-stdin foo.com
``` 

# Operating system and openconnect-gp version

openconnect-gp version:

```
OpenConnect version v7.08-240-gf9c36b4a
Using GnuTLS. Features present: HOTP software token, TOTP software token, System keys, DTLS, ESP
Supported protocols: anyconnect (default), nc, gp
```

Note that I've also tried this with version 7.07, with GnuTLS and with OpenSSL.

operating system

```
Linux localhost 4.9.68-x86_64-linode89 #1 SMP Mon Dec 11 16:09:29 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50
```

# GlobalProtect VPN information

```
Please enter your username and password
POST https://foo.com:444/ssl-vpn/login.esp
Attempting to connect to server 12.0.foo.foo:444
Connected to 12.0.foo.foo:444
SSL negotiation with foo.com
Connected to HTTPS on foo.com
> POST /ssl-vpn/login.esp HTTP/1.1
> Host: foo.com:444
> User-Agent: PAN GlobalProtect
> X-Pad: 0000000000000000000000000000000000000000000
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 149
> 
> jnlpReady=jnlpReady&ok=Login&direct=yes&clientVer=4100&prot=https:&server=foo.com&computer=localhost&user=surjikal&passwd=lol
Got HTTP response: HTTP/1.1 200 OK
Server: 
Date: Sat, 30 Dec 2017 07:01:17 GMT
Content-Type: application/xml; charset=UTF-8
Content-Length: 634
Connection: keep-alive
ETag: "bff8-1e6f-59cc41b3"
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-FRAME-OPTIONS: DENY
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
Set-Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9; secure; HttpOnly
HTTP body length:  (634)
< <?xml version="1.0" encoding="utf-8"?><jnlp><application-desc><argument>(null)</argument><argument>3e7092b1c16c4e0a3d29f1851269af53</argument><argument>d2ae2014ce75f0cf7ff43f173082aba0160ac0d3</argument><argument>Foo GP Gateway-N</argument><argument>surjikal</argument><argument>Foo Login Seq</argument><argument>vsys1</argument><argument>bar</argument><argument>(null)</argument><argument></argument><argument></argument><argument></argument><argument>tunnel</argument><argument>-1</argument><argument>4100</argument><argument></argument><argument></argument><argument></argument></application-desc></jnlp>
GlobalProtect login returned authentication-source=Foo Login Seq
POST https://foo.com:444/ssl-vpn/getconfig.esp
> POST /ssl-vpn/getconfig.esp HTTP/1.1
> Host: foo.com:444
> User-Agent: PAN GlobalProtect
> Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9
> X-Pad: 0000000000000000000000000000000000000000000000000000000
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 265
> 
> client-type=1&protocol-version=p1&app-version=3.0.1-10&os-version=linux-64&clientos=linux-64&hmac-algo=sha1%2cmd5&enc-algo=aes-128-cbc%2caes-256-cbc&authcookie=X&portal=Foo%20GP%20Gateway-N&user=surjikal&domain=bar
Got HTTP response: HTTP/1.1 200 OK
Server: 
Date: Sat, 30 Dec 2017 07:01:17 GMT
Content-Type: application/xml; charset=UTF-8
Content-Length: 1536
Connection: keep-alive
ETag: "bff2-1f2-59cc41b3"
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-FRAME-OPTIONS: DENY
HTTP body length:  (1536)
< 
< 	<response status="success">
< 		<need-tunnel>yes</need-tunnel>
< 		<ssl-tunnel-url>/ssl-tunnel-connect.sslvpn</ssl-tunnel-url>
< 		<portal>Foo GP Gateway-N</portal>
< 		<user>surjikal</user>
< 		<lifetime>2592000</lifetime>
< 		<timeout>10800</timeout>
< 		<disconnect-on-idle>108000</disconnect-on-idle>
< 		<bw-c2s>1000</bw-c2s>
< 		<bw-s2c>1000</bw-s2c>
< 		<gw-address>10.11.0.8</gw-address>
< 		<ip-address>10.11.70.249</ip-address>
< 		<netmask>255.255.255.255</netmask>
< 		<dns>
< 			<member>10.11.100.74</member>
< 			<member>10.11.100.65</member>
< 		</dns> 
< 		<wins>
< 		</wins> 
< 		<default-gateway>10.11.70.249</default-gateway>
< 		<mtu>0</mtu>
< 		<dns-suffix>
< 		</dns-suffix> 
< 		<no-direct-access-to-local-network>yes</no-direct-access-to-local-network>
< 		<access-routes>
< 			<member>0.0.0.0/0</member>
< 			<member>10.11.100.74/32</member>
< 			<member>10.11.100.65/32</member>
< 		</access-routes> 
< 		<ipsec>
< 			<udp-port>4501</udp-port>
< 			<ipsec-mode>esp-tunnel</ipsec-mode>
< 			<enc-algo>aes-128-cbc</enc-algo>
< 			<hmac-algo>sha1</hmac-algo>
< 			<c2s-spi>0x65EE9C2B</c2s-spi>
< 			<s2c-spi>0x55291B2D</s2c-spi>
< 			<akey-s2c>
< 				<bits>160</bits>
< 				<val>hex1</val>
< 			</akey-s2c> 
< 			<ekey-s2c>
< 				<bits>128</bits>
< 				<val>hex2</val>
< 			</ekey-s2c> 
< 			<akey-c2s>
< 				<bits>160</bits>
< 				<val>hex3</val>
< 			</akey-c2s> 
< 			<ekey-c2s>
< 				<bits>128</bits>
< 				<val>hex4</val>
< 			</ekey-c2s> 
< 		</ipsec> 
< 	</response>
Tunnel timeout (rekey interval) is 180 minutes.
Gateway address in config XML (10.11.0.8) differs from external gateway address (12.0.foo.foo).
TCP_INFO rcv mss 1460, snd mss 1460, adv mss 1460, pmtu 1500
No MTU received. Calculated 1410
POST https://foo.com:444/ssl-vpn/hipreportcheck.esp
> POST /ssl-vpn/hipreportcheck.esp HTTP/1.1
> Host: foo.com:444
> User-Agent: PAN GlobalProtect
> Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9
> X-Pad: 00000000000000000000000000000
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 227
> 
> authcookie=X&portal=Foo%20GP%20Gateway-N&user=surjikal&domain=bar&computer=localhost&client-ip=10.11.70.249&client-role=global-protect-full&md5=656043395701e0a21454be6b732535ae
Got HTTP response: HTTP/1.1 200 OK
Server: 
Date: Sat, 30 Dec 2017 07:01:17 GMT
Content-Type: application/xml; charset=UTF-8
Content-Length: 107
Connection: keep-alive
ETag: "bff5-441-59cc41b3"
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-FRAME-OPTIONS: DENY
HTTP body length:  (107)
< 
< 	<response status="success">
< 		<hip-report-needed>yes</hip-report-needed>
< 		<delay>0</delay>
< 	</response>
Gateway says HIP report submission is needed.
POST https://foo.com:444/ssl-vpn/hipreport.esp
> POST /ssl-vpn/hipreport.esp HTTP/1.1
> Host: foo.com:444
> User-Agent: PAN GlobalProtect
> Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9
> X-Pad: 000000000000
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 12148
> 
> authcookie=...snip...
Got HTTP response: HTTP/1.1 200 OK
Server: 
Date: Sat, 30 Dec 2017 07:01:17 GMT
Content-Type: application/xml; charset=UTF-8
Content-Length: 75
Connection: keep-alive
ETag: "bff4-3fa-59cc41b3"
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-FRAME-OPTIONS: DENY
HTTP body length:  (75)
< 
< 	<response status="success">
< 		<notification></notification>
< 	</response>
HIP report submitted successfully.
Send ESP probes
Connected as 10.11.70.249, using SSL
Failed to connect ESP tunnel; using HTTPS instead.
Connecting to HTTPS tunnel endpoint ...
> GET /ssl-tunnel-connect.sslvpn?authcookie=X&portal=Foo%20GP%20Gateway-N&user=surjikal&domain=bar HTTP/1.1
> 
Send GPST DPD/keepalive request
Got GPST DPD/keepalive response
POST https://foo.com:444/ssl-vpn/logout.esp
SSL negotiation with foo.com
Connected to HTTPS on foo.com
> POST /ssl-vpn/logout.esp HTTP/1.1
> Host: foo.com:444
> User-Agent: PAN GlobalProtect
> Cookie: PHPSESSID=ecfcee7c46a1ac046b50ab78a686c0e9
> X-Pad: 000000000000000000000000000000000000000000000000000000000
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 135
> 
> computer=localhost&authcookie=X&portal=Foo%20GP%20Gateway-N&user=surjikal&domain=bar
Got HTTP response: HTTP/1.1 200 OK
Server: 
Date: Sat, 30 Dec 2017 07:01:36 GMT
Content-Type: application/xml; charset=UTF-8
Content-Length: 177
Connection: keep-alive
ETag: "bff9-251-59cc41b3"
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-FRAME-OPTIONS: DENY
HTTP body length:  (177)
< 
< 	<response status="success">
< 		<portal>Foo GP Gateway-N</portal>
< 		<domain>bar</domain>
< 		<user>surjikal</user>
< 		<computer>localhost</computer>
< 	</response>
Logout successful
RTNETLINK answers: No such process
RTNETLINK answers: No such process
User cancelled (SIGINT); exiting.
```

My theories:

1. The internal gateway server is reachable only through the VPN, so when it tries to establish a tunnel, it fails because the VPN tunnel hasn't been established yet. However, I removed the line of code that does the internal gateway rewrite (basically reverted #25) but that did not fix it. I possibly did it wrong but my production openconnect instance is running on a version that was pulled in nov 2016, which would not include that fix...

```
Gateway address in config XML (10.11.0.8) differs from external gateway address (12.0.foo.foo).
```

2. The timeout for the ICMP reply is too low (I think it's 5 seconds currently).

3. ESP is misconfigured / disabled in the gateway, or the firewall is prevent ESP tunnels.

I don't know too much about the protocol or anything so my theories may be completely off.

**Comment by dlenski** [comment]
Thanks for the detailed issue report. Hmmm. Seems like you checked all the obvious things.

Are you sure you're not behind a firewall that blocks ESP?

1) Reverting #25 would almost certainly _not_ fix this problem, since #25 was the patch that enabled openconnect to interpret the `<gw-address>` in the intended way.

2) If the ESP gateway won't reply to a ping in 5 seconds, then ESP is not going to make your connection much faster :cry: … but did you try increasing the timeout here?
 https://github.com/dlenski/openconnect/blob/HEAD/gpst.c#L841-L842
(It might be worth firing up Wireshark on your Linux box as well to verify that you're really not receiving *any* ESP return packets.)

3) Are you _sure_ that your VPN will connect successfully via ESP from one of the official clients? There have been previous issue reports of VPNs which say that they support ESP… but simply will not connect to ESP, no matter what. (You may need to fire up Wireshark or tcpdump to verify that it's sending ESP packets, not HTTPS packets, if the official client doesn't tell you how it's connected.)

Bottom line is you should verify whether ESP works with an official client first!

**Comment by surjikal** [comment]
Thank you for your reply! I am indeed not sure if ESP is even supported. I
can definitely test using wireshark on a Mac or windows box using the
official client.

How can I tell if I'm getting an ESP connection in wireshark? Should I
filter the traffic by the external IP (12.0.foo.foo) and just look for ESP?
Sorry I'm quite familiar with devops stuff but VPN connections are alien
technology to me.

**Comment by dlenski** [comment]
> How can I tell if I'm getting an ESP connection in wireshark? Should I filter the traffic by the external IP (12.0.foo.foo) and just look for ESP?

Yeah, even just looking for UDP traffic to port 4501 of the external IP should do it! **Make sure you're behind the same firewall (or lack thereof) as you are when connecting from your Linux system.**

> 			<udp-port>4501</udp-port>

If the traffic is all going over HTTPS, on the other hand, then you'll see tons of TCP traffic on port 443.

_Some_ versions of the official clients tell you whether connected via HTTPS or ESP, or you can tell from the log files… but Wireshark is probably easiest.

> Sorry I'm quite familiar with devops stuff but VPN connections are alien technology to me.

I wish I hadn't had to learn all this. :sob:

**Comment by surjikal** [comment]
I don't think I can reproduce the network environment of my linux server unfortunately. But I think I may not need to....

![panel](https://i.imgur.com/wmqwAG8.png)

I also used wireshark and all the traffic was over TCP (was port 444, which is the port of the external gateway).

And then just in case, I set the logs to "dump", collected them and read through `PanGPS.log`, and saw this:

```
P7931-T133895 Dec 30 19:05:56:760407 Debug( 162): Trying to do ipsec connection to 12.0.foo.foo[4501]
P7931-T133895 Dec 30 19:05:56:760415 Dump ( 695): 12.0.foo.foo is not ipv6
P7931-T133667 Dec 30 19:05:56:760429 Debug(1374): Route change message RTM_NEWADDR: address being added to iface gpd0: 10.11.70.22
P7931-T133667 Dec 30 19:05:56:760446 Dump (1365): Route change message RTM_ADD: Add Route 10.11.70.22
P7931-T130839 Dec 30 19:05:56:760455 Debug(4246): NetworkConnectionMonitorThread: route change detected. Wait for 3 seconds.
P7931-T133895 Dec 30 19:05:56:760470 Dump ( 479): UDP socket set SO_SNDBUF size 1232896
P7931-T133895 Dec 30 19:05:56:760477 Dump ( 486): UDP socket set SO_RCVBUF size 1232896
P7931-T133667 Dec 30 19:05:56:760485 Debug(1383): Route change message RTM_IFINFO: iface status change, gpd0 up
P7931-T133895 Dec 30 19:05:56:760493 Dump ( 490): Connect timeout is 60
P7931-T133895 Dec 30 19:05:56:761393 Info ( 174): Connected to: 12.0.foo.foo[4501], Sending keep alive to ipsec socket...
P7931-T133895 Dec 30 19:06:02:964979 Info ( 217): failed to receive keep alive
P7931-T133895 Dec 30 19:06:02:964993 Debug( 226): Disconnect udp socket
P7931-T133895 Dec 30 19:06:02:965038 Info ( 295): Connecting to 12.0.foo.foo failed
P7931-T133895 Dec 30 19:06:02:965043 Info ( 224): Start vpn do_connect() failed
P7931-T133895 Dec 30 19:06:02:965046 Debug( 268): do_disconnect is called in VPN stop
P7931-T133895 Dec 30 19:06:02:965049 Debug( 271): unsetnetwork is called in vpn stop
```

So things are not looking too good...

**Comment by dlenski** [comment]
> It could be my personal network but I'm having trouble getting ESP to work on 2 different linux servers (one is completely open) as well so... hmm. I guess I'll have to go through IT and see if I can get them to look into this.

Yeah, you've demonstrated it clearly. It seems ESP is borked on the server side of the VPN. Not the first time I've seen it, I'm afraid.

Hopefully you can get IT to fix it, but in my experience "don't hold your breath" is a common outcome.

**Comment by dlenski** [comment]
By the way, @surjikal, I also saw that you're using the newly-added HIP submission support. Is that working for you?

I haven't got confirmation from anyone (besides myself :grin:) that it works in its current form, so it'd be great to get an additional data point or feedback.

**Comment by surjikal** [comment]
It seems to be working great! I'm pretty sure I started using it to see if it would somehow help with the ESP stuff. All I had to do was fill in the `USER` and `DOMAIN` variables, which I got from the logs of the native client (though I'm sure you can get them from `--dump`).

**Comment by dlenski** [comment]
> I filled in the USER and DOMAIN variables manually. I got them from the native client logs, but I'm sure you can also get them from --dump.

Ah, you shouldn't need to fill those in manually. [The script](https://github.com/dlenski/openconnect/blob/HEAD/hipreport.sh) should automatically parse them out of the cookie:

```bash
# Extract username and domain from cookie
USER=$(echo "$COOKIE" | sed -rn 's/(.+&|^)user=([^&]+)(&.+|$)/\2/p')
DOMAIN=$(echo "$COOKIE" | sed -rn 's/(.+&|^)domain=([^&]+)(&.+|$)/\2/p')
```

Is that not working?

> One minor thing, it wasn't immediately obvious that the other variables would be filled in by the script invocation (but I figured it out after reading through it).

Yeah, I should tweak the comment to explain that `openconnect` itself will call the script with the required arguments.

**Comment by surjikal** [comment]
Well it looks like I somehow used an old version of the script then! The head looks like this:

```bash
#!/bin/bash
COOKIE=
COMPUTER=
IP=
MD5=
NOW=$(date +'%m/%d/%Y %H:%M:%S')
USER="<I added this>"
DOMAIN="<I added this too>"

while [ "$1" ]; do
```

**Comment by dlenski** [comment]
Right, you're using the initial version. I've added more comments explaining how it works and tries to automatically extract the `USER` and `DOMAIN` values.

https://github.com/dlenski/openconnect/commits/HEAD/hipreport.sh
